pkg/resources/share.go

```diff
@@ -112,92 +123,127 @@ func setAccounts(d *schema.ResourceData, meta interface{}) error {
+		Add: &sdk.ShareAdd{
+			Accounts: accounts,
+		},
+	})
+	return err
 }
 
 // ReadShare implements schema.ReadFunc.
 func ReadShare(d *schema.ResourceData, meta interface{}) error {
 	db := meta.(*sql.DB)
-	id := d.Id()
-
-	stmt := snowflake.NewShareBuilder(id).Show()
-	row := snowflake.QueryRow(db, stmt)
+	id := sdk.NewAccountObjectIdentifier(d.Id())
+	client := sdk.NewClientFromDB(db)
+	ctx := context.Background()
 
-	s, err := snowflake.ScanShare(row)
-	if errors.Is(err, sql.ErrNoRows) {
-		// If not found, mark resource to be removed from state file during apply or refresh
-		log.Printf("[DEBUG] share (%s) not found", d.Id())
-		d.SetId("")
-		return nil
-	}
+	share, err := client.Shares.ShowByID(ctx, id)
 	if err != nil {
-		return err
+		return fmt.Errorf("error reading share err = %w", err)
 	}
-
-	if err := d.Set("name", StripAccountFromName(s.Name.String)); err != nil {
+	if err := d.Set("name", share.Name.Name()); err != nil {
 		return err
 	}
-	if err := d.Set("comment", s.Comment.String); err != nil {
+	if err := d.Set("comment", share.Comment); err != nil {
 		return err
 	}
+	accounts := make([]string, len(share.To))
+	for i, accountIdentifier := range share.To {
+		accounts[i] = accountIdentifier.Name()
+	}
 
-	accs := strings.FieldsFunc(s.To.String, func(c rune) bool { return c == ',' })
-	err = d.Set("accounts", accs)
+	if err := d.Set("accounts", accounts); err != nil {
+		return err
+	}
 
 	return err
 }
 
+func accountIdentifiersFromSlice(accounts []string) []sdk.AccountIdentifier {
+	accountIdentifiers := make([]sdk.AccountIdentifier, len(accounts))
+	for i, account := range accounts {
+		parts := strings.Split(account, ".")
+		orgName := parts[0]
+		accountName := parts[1]
+		accountIdentifiers[i] = sdk.NewAccountIdentifier(orgName, accountName)
+	}
+	return accountIdentifiers
+}
+
 // UpdateShare implements schema.UpdateFunc.
 func UpdateShare(d *schema.ResourceData, meta interface{}) error {
-	// Change the accounts first - this is a special case and won't work using the generic method
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	ctx := context.Background()
 	if d.HasChange("accounts") {
-		if err := setAccounts(d, meta); err != nil {
-			return err
+		o, n := d.GetChange("accounts")
+		oldAccounts := expandStringList(o.([]interface{}))
+		newAccounts := expandStringList(n.([]interface{}))
+		if len(newAccounts) == 0 {
+			accountIdentifiers := accountIdentifiersFromSlice(oldAccounts)
+			err := client.Shares.Alter(ctx, sdk.NewAccountObjectIdentifier(d.Id()), &sdk.ShareAlterOptions{
+				Remove: &sdk.ShareRemove{
+					Accounts: accountIdentifiers,
+				},
+			})
+			if err != nil {
+				return fmt.Errorf("error removing accounts from share err = %w", err)
+			}
+		} else {
+			accountIdentifiers := accountIdentifiersFromSlice(newAccounts)
+			err := setShareAccounts(ctx, client, sdk.NewAccountObjectIdentifier(d.Id()), accountIdentifiers)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
If an account has been removed in the terraform config then this won't remove it from the share, this only adds accounts to the share. Correct? I think there's a bug there.

Reply:
according to documentation set accounts does the following:
> Specifies the account(s) to replace all previous accounts with which the share was shared. T
so yes it will remove accounts. unless the list is zero length, in which case i would have it 
we already have a test that goes from 1 account down to 0 accounts. I added a test that goes from 2 -> 1 also. Noticed there was an issue with the ordering of the result (fixed), but it does correctly remove accounts